Update the example to the new driver API

AndrzejKurek opened the pull request:
The key slot numbers used in PSA currently map directly to the hardware slots.
Because of this, as slot 0 is reserved in PSA, slot 1 is used for private key instead.
Change direct low-level driver calls to calls to the PSA API.
Remove the public key exporting during private key generation test, since this functionality is not present anymore.
The storage is purged after the first key generation test for two reasons:
first - so that the same key slot can be reused and second - so that key
registration can be tested as a second way to have a key in a PSA slot.

This PR is based on:
- https://github.com/ARMmbed/mbed-os-atecc608a/pull/7 - driver changes;
- https://github.com/ARMmbed/mbed-crypto/pull/157 - key foundation;
- https://github.com/ARMmbed/mbed-crypto/pull/183 - key registration;
- https://github.com/ARMmbed/mbed-crypto/pull/174 - hooks.

Patater committed:
Enable SE support in Mbed Crypto

Patater commented:
Could we please rebase this to apply fixups and remove the merge commits?

k-stachowiak commented:
I have tested this, with the `mbed-os-atecc608a.lib` containing:

`https://github.com/AndrzejKurek/mbed-os-atecc608a/#7cc48d30f16474fde86b12d052a6384a14926205`,

Which currently is the head of Andrzej's `key-registration-api` branch.

I confirm it works and passes all the tests on K64F, built with GCC_ARM.

AndrzejKurek commented:
@Patater done.

k-stachowiak commented:
I have repeated the test after the rebase and the tests still pass.

Patater commented on atecc608a/main.c:
```diff
@@ -26,7 +26,7 @@
 #include "atca_helpers.h"
 #include "atecc508a_config_dev.h"
-
+#include "psa_crypto_storage.h"
```
This is a private interface. Can we use `mbed_psa_reboot_and_request_new_security_state(PSA_LIFECYCLE_ASSEMBLY_AND_TEST);` instead? This comes from "psa/lifecycle.h"

k-stachowiak replied:
I experimented with the function you suggested, and It works. Upon Andrzej's request I manually ran the test few times in a row, and they still pass.

Temporary removal of the reset code caused the tests to fail, so we made sure this step is necessary and the solution is valid.

Patater commented on atecc608a/main.c:
```diff
@@ -76,18 +76,32 @@
+    }
+    /* Purge the transaction file. */
+    psa_crypto_stop_transaction();
+}
```
This function is yucky to put into an example, which should be more whitebox style (what users should actually be recommended to do and can use).

AndrzejKurek replied:
We have to compromise between showing an example and testing somehow. Purging the storage is needed.

Patater replied:
`mbed_psa_reboot_and_request_new_security_state(PSA_LIFECYCLE_ASSEMBLY_AND_TEST); ` will purge storage in a slightly more acceptable way

Patater commented on atecc608a/main.c:
```diff
@@ -76,18 +76,32 @@
 /* Data used by tests */
-psa_key_slot_number_t atecc608a_private_key_slot = 0;
+psa_key_slot_number_t atecc608a_private_key_slot = 1;
 psa_key_slot_number_t atecc608a_public_key_slot = 9;
```
If we got rid of these, we'd have to have a driver that implemented "allocate", right? Since this is non-ideal, we should leave a TODO comment to tell users "don't do this" and to remind us to fix and file an issue.

AndrzejKurek replied:
In the example - yes. We would still need to have information about the hardware configuration somewhere. Ideally - the driver could read it in runtime and allocate a key, but that's a lot of work, and it's flaky (unless we make it bloated and check **a lot** of stuff).
Oh, and having the hardware configured to have a key in a given slot still wouldn't mean that it's there, so I still see a use case for hardcoding it.

Patater commented on atecc608a/main.c:
```diff
@@ -410,9 +342,41 @@ psa_status_t test_hash_sha256()
+void setup_key_attributes(psa_key_attributes_t *attributes,
+                          psa_key_slot_number_t slot, int is_private)
+{
+    psa_set_key_slot_number(attributes, slot);
```
Ideally we wouldn't hard code the slot number. We shouldn't encourage hard coding the slot number in our examples.

AndrzejKurek replied:
There's a use case for that - if the user wants to use a certain pre-provisioned key, he/she should be able to point to it somehow.

Patater replied:
Yes. We should probably split those examples at some point.

Patater commented on atecc608a/main.c:
```diff
@@ -194,180 +208,98 @@ psa_status_t test_psa_import_verify()
 
-/* Test that a public key generated while generating a private key can
- * be imported. */
-psa_status_t test_generate_import()
```
Why can't we test that a public key generated while generating a private key can be imported anymore?

AndrzejKurek replied:
Last time I checked, the psa key generation function discarded the returned public key, as it would require implementing a storage. Not sure if it's the same in the current state though.

Patater replied:
Can we call export later instead of using the public key returned by the driver's generate?

AndrzejKurek replied:
Yes, we're performing such tests.

k-stachowiak committed:
Rely on the public PSA API to reset security state

AndrzejKurek commented:
Please note that before this PR is merged, the `mbed-os-atecc608a.lib` file has to point to the result of merging https://github.com/ARMmbed/mbed-os-atecc608a/pull/7.

gilles-peskine-arm commented on atecc608a/main.c:
```diff
@@ -194,180 +195,98 @@ psa_status_t test_psa_import_verify()
+    psa_close_key(verify_handle);
     printf("test_psa_import_verify succesful!\n");
 exit:
     return status;
```
There's a handle leak if an error occurs after opening `private_handle`. The `psa_close_key` calls should be after `exit:`. Note that `psa_close_key(0)` is explicitly defined to do nothing, so it's ok to call `psa_close_key` unconditionally if you initialize handle variables to 0.

The same pattern should be corrected in other functions as well.

gilles-peskine-arm commented on atecc608a/main.c:
```diff
@@ -77,17 +77,18 @@
-    key_bits = 256,
+    keypair_type = PSA_KEY_TYPE_ECC_KEY_PAIR(PSA_ECC_CURVE_SECP256R1),
+    priv_bits = PSA_BYTES_TO_BITS(32),
+    pub_bits = PSA_BYTES_TO_BITS(64),
```
No, `pub_bits` should be equal to `priv_bits`, and in fact there should be a single constant. A private key and the corresponding public key have the same bit size. The bit size is a characteristic of the mathematical object, not the size of the representation which depends on the representation format. For a SECP256R1 key, the key size is 256 bits.

gilles-peskine-arm commented on atecc608a/main.c:
```diff
@@ -443,7 +406,7 @@ void print_device_info()
     atecc608a_print_config_zone();
     atecc608a_print_locked_zones();
-    printf("\nPrivate key slot in use: %lu, public: %lu\n",
+    printf("\nPrivate key slot in use: %llu, public: %llu\n",
```
Minor: `atecc608a_private_key_slot` and `atecc608a_public_key_slot` are `uint64_t`, not `unsigned long long`. This is minor because this code is very unlikely to run on a platform where these two types are different. However, it would be better to be consistent. Note that since slot numbers in this particular driver (and in any driver based on the same model where the secure element has a static array of slots) are small, you could also use `%u` and `(unsigned) xxx_slot`.

gilles-peskine-arm commented on atecc608a/main.c:
```diff
@@ -477,10 +440,12 @@ bool interactive_loop()
     } else if (strcmp(command, "test") == 0) {
         run_tests();
-    } else if (strncmp(command, "generate_private", strlen("generate_private") - 1) == 0) {
+    } else if (strncmp(command, "generate_private",
```
Why does this match not only `"generate_private"` but also e.g. `"generate_private_wibble"`?

AndrzejKurek replied:
To match `"generate_private=5"` too, if an argument is to be supplied.

gilles-peskine-arm replied:
Ok. It would be cleaner to only allow `"generate_private\0"` and `"generate_private="`, but that's not a problem. This is not the part of the example where we demonstrate how to use the API, so it's ok to take shortcuts here.

gilles-peskine-arm commented on atecc608a/main.c:
```diff
@@ -223,10 +223,11 @@ psa_status_t test_psa_import_verify(psa_key_handle_t *private_handle,
+
 exit:
+    psa_close_key(*private_handle);
+    psa_close_key(verify_handle);
```
`verify_handle` may be uninitialized if there was an error before the call to `psa_open_key` that uses it. Initialize it to 0.

gilles-peskine-arm commented on atecc608a/main.c:
```diff
@@ -223,10 +223,11 @@ psa_status_t test_psa_import_verify(psa_key_handle_t *private_handle,
     printf("test_psa_import_verify succesful!\n");
+
 exit:
+    psa_close_key(*private_handle);
```
`*private_handle` is guaranteed to be initialized because nothing in this function can fail before the call to `psa_open_key` that uses it. But this is very fragile. Please initialize `*private_handle` to 0 before doing anything that can fail.

gilles-peskine-arm reviewed (changes requested):
The handle initialization is not right (I only checked one function but please check them all). Otherwise LGTM.

Commit:
Change the format of key slot printing

Patater commented:
The commit "Revert incorredt differentiation between priv..." should be applied as a fixup to the commit that caused the trouble. Does the example work after this reversion? The key size change was made to get the example working, but maybe we fixed that in another way.

gilles-peskine-arm commented:
Please avoid fixups for nontrivial issues found during a review.

I haven't checked how the code could work with incorrect sizes, but I suspect that it did because the sizes were never actually used in any calculation. For an ECC key, the key type determines everything and the bit size is redundant.

Patater commented:
Please avoid adding dead end wrong solutions to the permanent history. Keeps mistakes made on your own branches within your own branches. This issue is dumb enough that we don't even need a code comment explaining our past failed attempt. The key size is always 256 bits.

Commit:
Update the example to the new driver API.
The key slot numbers used in PSA currently map directly to the hardware slots.
Because of this, as slot 0 is reserved in PSA, slot 1 is used for private key
instead.
Change direct low-level driver calls to calls to the PSA API.
Remove the public key exporting during private key generation test, since this
functionality is not present anymore.
The storage is purged after the first key generation test for two reasons:
first - so that the same key slot can be reused and second - so that key
registration can be tested as a second way to have a key in a PSA slot.
Update the pointer to the new driver library in mbed-os-atecc608a.lib

AndrzejKurek commented:
Cleared the history and addressed all of the comments. @gilles-peskine-arm, @Patater - please re-review.

Patater commented:
What testing was done with this?

Patater commented:
Here's what I get

```
$ mbed compile -m K64F -t GCC_ARM --flash --sterm
...
Private key slot in use: 1, public: 9
Running tests...
test_hash_sha256 succesful!
test_sign_verify succesful!
test_export_import succesful!
test_psa_import_verify succesful!
assertion failed at ./main.c:392 (actual=109 expected=0)


Available commands:
 - info - print configuration information;
 - test - run all tests on the device;
 - exit - exit the interactive loop;
 - generate_private[=%d] - generate a private key in a given slot (0-15),
                           default slot - 0.
 - generate_public=%d_%d - generate a public key in a given slot
```

Patater commented:
After running `lock_data`, I was able to proceed.

```
test
Running tests...
test_hash_sha256 succesful!
test_sign_verify succesful!
test_export_import succesful!
test_psa_import_verify succesful!
test_write_read_slot succesful!
```

Patater reviewed (approved):
LGTM

Patater commented:
Tested on K64F+HAS_CRYPTOSHIELD

AndrzejKurek commented:
This was also tested by me on K64F with ATECC608A connected - I've run the test suites in succession and the tests passed.
